src/librustc/hir/pat_util.rs

```diff
@@ -21,6 +21,28 @@ use std::cell::RefCell;
 
 pub type PatIdMap = FnvHashMap<ast::Name, ast::NodeId>;
 
+#[derive(Clone, Copy)]
+pub struct AjustPos {
+    gap_pos: usize,
+    gap_len: usize,
+}
+
+impl FnOnce<(usize,)> for AjustPos {
+    type Output = usize;
+    extern "rust-call" fn call_once(self, (i,): (usize,)) -> usize {
+        if i < self.gap_pos { i } else { i + self.gap_len }
+    }
+}
+
+// Returns a functional object used to adjust tuple pattern indexes. Example: for 5-tuple and
+// pattern (a, b, .., c) expected_len is 5, actual_len is 3 and gap_pos is Some(2).
+pub fn pat_adjust_pos(expected_len: usize, actual_len: usize, gap_pos: Option<usize>) -> AjustPos {
```

Comment:
Not that it matters, but I believe that every instance of this is in fact being used to modify indices resulting from an `Enumerate`. I could imagine making this an iterator adapter instead, might be cleaner and would mean that you never have an "unadjusted" index (e.g., `for (i, pat) in pats.iter().enumerate_and_adjust(..)`).